File: src/services/redditClient.ts
import { RedditConfig, RedditAuthResponse, RedditPost, RedditUser, SocialSentiment } from '../types/Reddit';

export class RedditClient {
  private config: RedditConfig;
  private accessToken: string | null = null;
  private tokenExpiry: Date | null = null;

  // Hype language detection terms for shopping/retail context
  private readonly hypeTerms = new Set([
    'amazing', 'incredible', 'unbelievable', 'steal', 'deal', 'bargain',
    'must have', 'love it', 'obsessed', 'game changer', 'life changing',
    'worth every penny', 'best purchase', 'highly recommend', 'amazing deal',
    'unbeatable price', 'incredible value', 'fantastic', 'outstanding',
    'excellent', 'perfect', 'awesome', 'brilliant', 'superb', 'terrific',
    'fabulous', 'wonderful', 'marvelous', 'splendid', 'magnificent',
    'extraordinary', 'phenomenal', 'exceptional', 'remarkable', 'impressive'
  ]);

  constructor(config: RedditConfig) {
    this.config = config;
  }

  async getSocialSentimentAsync(product: string, retailer?: string): Promise<SocialSentiment | null> {
    try {
      console.log(`Fetching Reddit data for ${product}${retailer ? ` at ${retailer}` : ''}`);

      // Authenticate first
      await this.ensureAuthenticated();

      // Search across configured subreddits
      const mentions = new List<RedditPost>();
      for (const subreddit of this.config.subreddits) {
        try {
          const posts = await this.searchSubredditAsync(subreddit, product, retailer);
          mentions.addRange(posts);

          // Rate limiting (60 requests/minute for OAuth)
          await this.delay(1000);
        } catch (error) {
          console.error(`Error searching r/${subreddit} for ${product}:`, error);
        }
      }

      if (mentions.length === 0) {
        console.log(`No Reddit mentions found for ${product}`);
        return {
          product,
          retailer: retailer || '',
          windowStartUtc: new Date(Date.now() - 24 * 60 * 60 * 1000).toISOString(),
          messageCount: 0,
          hypeLexiconDensity: 0.0,
          newAccountRatio: 0.0,
          sentimentScore: 0.0,
          trending: false
        };
      }

      // Fetch author details for account age analysis
      const authorDetails = await this.getAuthorDetailsAsync(
        [...new Set(mentions.map(m => m.author))].slice(0, 50)
      );

      // Calculate metrics
      const hypeCount = mentions.filter(m =>
        this.containsHypeLanguage(m.title + ' ' + m.selftext)
      ).length;
      const hypeDensity = mentions.length > 0 ? hypeCount / mentions.length : 0.0;

      const newAccountCount = authorDetails.filter(a =>
        this.isNewAccount(a, 90)
      ).length;
      const newAccountRatio = authorDetails.length > 0 ? newAccountCount / authorDetails.length : 0.0;

      // Calculate sentiment score (simplified - could be enhanced with NLP)
      const sentimentScore = this.calculateSentimentScore(mentions);

      const trending = this.isTrending(mentions);

      console.log(`Reddit sentiment for ${product}: ${mentions.length} mentions, ${Math.round(hypeDensity * 100)}% hype, ${Math.round(newAccountRatio * 100)}% new accounts`);

      return {
        product,
        retailer: retailer || '',
        windowStartUtc: new Date(Date.now() - 24 * 60 * 60 * 1000).toISOString(),
        messageCount: mentions.length,
        hypeLexiconDensity: hypeDensity,
        newAccountRatio: newAccountRatio,
        sentimentScore,
        trending
      };
    } catch (error) {
      console.error(`Failed to fetch Reddit data for ${product}:`, error);
      return null;
    }
  }

  private async ensureAuthenticated(): Promise<void> {
    if (this.accessToken && this.tokenExpiry && new Date() < this.tokenExpiry) {
      return;
    }

    console.log('Authenticating with Reddit API');

    const authString = btoa(`${this.config.clientId}:${this.config.clientSecret}`);
    const response = await fetch('https://www.reddit.com/api/v1/access_token', {
      method: 'POST',
      headers: {
        'Authorization': `Basic ${authString}`,
        'User-Agent': this.config.userAgent,
        'Content-Type': 'application/x-www-form-urlencoded'
      },
      body: new URLSearchParams({
        grant_type: 'client_credentials'
      })
    });

    if (!response.ok) {
      throw new Error(`Reddit auth failed: ${response.status} ${response.statusText}`);
    }

    const authResp: RedditAuthResponse = await response.json();

    if (!authResp.access_token) {
      throw new Error('Failed to get access token from Reddit');
    }

    this.accessToken = authResp.access_token;
    this.tokenExpiry = new Date(Date.now() + (authResp.expires_in - 60) * 1000); // 60s buffer

    console.log('Reddit authentication successful');
  }

  private async searchSubredditAsync(subreddit: string, product: string, retailer?: string): Promise<RedditPost[]> {
    const posts: RedditPost[] = [];

    // Search query: product name (with retailer if specified)
    let query = product;
    if (retailer) {
      query += ` ${retailer}`;
    }

    const url = `https://oauth.reddit.com/r/${subreddit}/search?q=${encodeURIComponent(query)}&restrict_sr=1&sort=new&t=day&limit=${this.config.searchLimitPerSubreddit}`;

    const response = await fetch(url, {
      headers: {
        'Authorization': `Bearer ${this.accessToken}`,
        'User-Agent': this.config.userAgent
      }
    });

    if (!response.ok) {
      console.warn(`Reddit search failed for r/${subreddit}: ${response.status}`);
      return posts;
    }

    const data = await response.json();

    if (!data.data?.children) {
      return posts;
    }

    for (const child of data.data.children) {
      const postData = child.data;

      const post: RedditPost = {
        id: postData.id || '',
        author: postData.author || '',
        title: postData.title || '',
        selftext: postData.selftext || '',
        created_utc: postData.created_utc || 0,
        score: postData.score || 0,
        num_comments: postData.num_comments || 0,
        subreddit,
        url: postData.url || '',
        permalink: postData.permalink || ''
      };

      // Verify product/retailer is actually mentioned (Reddit search can be loose)
      const content = (post.title + ' ' + post.selftext).toLowerCase();
      const searchTerms = [product.toLowerCase()];
      if (retailer) {
        searchTerms.push(retailer.toLowerCase());
      }

      if (searchTerms.some(term => content.includes(term))) {
        posts.push(post);
      }
    }

    console.log(`Found ${posts.length} posts mentioning ${product} in r/${subreddit}`);

    return posts;
  }

  private async getAuthorDetailsAsync(authors: string[]): Promise<RedditUser[]> {
    const users: RedditUser[] = [];

    for (const author of authors) {
      try {
        // Skip deleted/suspended accounts
        if (author.startsWith('[deleted]') || author.startsWith('[removed]')) {
          continue;
        }

        const response = await fetch(`https://oauth.reddit.com/user/${author}/about`, {
          headers: {
            'Authorization': `Bearer ${this.accessToken}`,
            'User-Agent': this.config.userAgent
          }
        });

        if (!response.ok) {
          console.warn(`Failed to get details for u/${author}: ${response.status}`);
          continue;
        }

        const data = await response.json();

        if (data.data) {
          users.push(data.data);
        }

        // Rate limiting (60 req/min)
        await this.delay(1000);
      } catch (error) {
        console.warn(`Error fetching details for u/${author}:`, error);
      }
    }

    return users;
  }

  private containsHypeLanguage(text: string): boolean {
    const lower = text.toLowerCase();
    return Array.from(this.hypeTerms).some(term => lower.includes(term));
  }

  private isNewAccount(user: RedditUser, dayThreshold: number): boolean {
    const accountAge = (Date.now() / 1000) - user.created_utc;
    const days = accountAge / (24 * 60 * 60);
    return days < dayThreshold;
  }

  private calculateSentimentScore(posts: RedditPost[]): number {
    // Simplified sentiment analysis based on upvotes and comments
    const totalScore = posts.reduce((sum, post) => sum + post.score, 0);
    const totalComments = posts.reduce((sum, post) => sum + post.num_comments, 0);

    // Normalize to 0-1 scale
    const avgScore = posts.length > 0 ? totalScore / posts.length : 0;
    const avgComments = posts.length > 0 ? totalComments / posts.length : 0;

    // Simple sentiment score: higher scores and comments = more positive sentiment
    const score = Math.min(1, (avgScore / 50 + avgComments / 10) / 2);

    return Math.max(0, score);
  }

  private isTrending(posts: RedditPost[]): boolean {
    // Consider trending if recent posts have high engagement
    const recentPosts = posts.filter(post =>
      (Date.now() / 1000) - post.created_utc < 24 * 60 * 60 // Last 24 hours
    );

    if (recentPosts.length === 0) return false;

    const avgScore = recentPosts.reduce((sum, post) => sum + post.score, 0) / recentPosts.length;
    const avgComments = recentPosts.reduce((sum, post) => sum + post.num_comments, 0) / recentPosts.length;

    // Trending threshold: avg score > 10 and avg comments > 5
    return avgScore > 10 && avgComments > 5;
  }

  private delay(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}

// Factory function for easy configuration
export function createRedditClient(config?: Partial<RedditConfig>): RedditClient {
  const defaultConfig: RedditConfig = {
    clientId: import.meta.env.VITE_REDDIT_CLIENT_ID || '',
    clientSecret: import.meta.env.VITE_REDDIT_CLIENT_SECRET || '',
    userAgent: 'VeritasAlpha-Deals/1.0',
    subreddits: ['deals', 'shopping', 'frugal', 'buildapcsales'],
    searchLimitPerSubreddit: 50
  };

  return new RedditClient({ ...defaultConfig, ...config });
}

File: VeritasAlpha/Ingestion/RedditClient.cs
// =======================
// Reddit Integration for Veritas Alpha
// =======================
// Replace SocialClientStub with this implementation
// Requires Reddit app credentials: https://www.reddit.com/prefs/apps
// =======================

using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using VeritasAlpha.Core;

namespace VeritasAlpha.Ingestion
{
    // =======================
    // Configuration Model
    // =======================
    public sealed class RedditConfig
    {
        public string ClientId { get; set; } = string.Empty;
        public string ClientSecret { get; set; } = string.Empty;
        public string UserAgent { get; set; } = "VeritasAlpha/1.0";
        public string[] Subreddits { get; set; } = new[] { "stocks", "investing", "wallstreetbets" };
        public int SearchLimitPerSubreddit { get; set; } = 100;
    }

    // =======================
    // Reddit API Models
    // =======================
    internal sealed class RedditAuthResponse
    {
        public string access_token { get; set; } = string.Empty;
        public string token_type { get; set; } = string.Empty;
        public int expires_in { get; set; }
        public string scope { get; set; } = string.Empty;
    }

    internal sealed class RedditPost
    {
        public string Id { get; set; } = string.Empty;
        public string Author { get; set; } = string.Empty;
        public string Title { get; set; } = string.Empty;
        public string Selftext { get; set; } = string.Empty;
        public long Created { get; set; }
        public int Score { get; set; }
        public int NumComments { get; set; }
        public string Subreddit { get; set; } = string.Empty;
    }

    internal sealed class RedditUser
    {
        public string Name { get; set; } = string.Empty;
        public long Created { get; set; }
        public int LinkKarma { get; set; }
        public int CommentKarma { get; set; }
    }

    // =======================
    // Reddit Client Implementation
    // =======================
    public sealed class RedditClient : ISocialClient, IDisposable
    {
        private readonly HttpClient _http;
        private readonly RedditConfig _config;
        private readonly ILogger<RedditClient> _log;
        private readonly HashSet<string> _hypeTerms;
        
        private string? _accessToken;
        private DateTime _tokenExpiry = DateTime.MinValue;
        private readonly SemaphoreSlim _authLock = new(1, 1);

        public RedditClient(HttpClient http, RedditConfig config, ILogger<RedditClient> log)
        {
            _http = http;
            _config = config;
            _log = log;

            // Hype language detection terms
            _hypeTerms = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
            {
                "moon", "🚀", "rocket", "lambo", "diamond hands", "💎",
                "squeeze", "short squeeze", "ape", "hodl", "yolo", 
                "tendies", "to the moon", "10x", "100x", "moonshot",
                "🌙", "🦍", "pump", "💰", "🤑"
            };
        }

        // =======================
        // Main Interface Method
        // =======================
        public async Task<SocialSummary?> GetRedditSummaryAsync(string ticker)
        {
            try
            {
                _log.LogInformation("Fetching Reddit data for {Ticker}", ticker);

                // Authenticate first
                await EnsureAuthenticatedAsync();

                // Search across configured subreddits
                var mentions = new List<RedditPost>();
                foreach (var subreddit in _config.Subreddits)
                {
                    try
                    {
                        var posts = await SearchSubredditAsync(subreddit, ticker);
                        mentions.AddRange(posts);
                        
                        // Rate limiting (60 requests/minute for OAuth)
                        await Task.Delay(TimeSpan.FromSeconds(1));
                    }
                    catch (Exception ex)
                    {
                        _log.LogError(ex, "Error searching r/{Subreddit} for {Ticker}", subreddit, ticker);
                    }
                }

                if (mentions.Count == 0)
                {
                    _log.LogInformation("No Reddit mentions found for {Ticker}", ticker);
                    return new SocialSummary
                    {
                        Ticker = ticker,
                        WindowStartUtc = DateTime.UtcNow.AddHours(-24),
                        MessageCount = 0,
                        HypeLexiconDensity = 0.0,
                        NewAccountRatio = 0.0
                    };
                }

                // Fetch author details for account age analysis
                var authorDetails = await GetAuthorDetailsAsync(mentions.Select(m => m.Author).Distinct().Take(50));

                // Calculate metrics
                var hypeCount = mentions.Count(m => ContainsHypeLanguage(m.Title + " " + m.Selftext));
                var hypeDensity = (double)hypeCount / mentions.Count;

                var newAccountCount = authorDetails.Count(a => IsNewAccount(a, dayThreshold: 90));
                var newAccountRatio = authorDetails.Count > 0 ? (double)newAccountCount / authorDetails.Count : 0.0;

                _log.LogInformation("Reddit summary for {Ticker}: {Count} mentions, {Hype:P0} hype, {New:P0} new accounts",
                    ticker, mentions.Count, hypeDensity, newAccountRatio);

                return new SocialSummary
                {
                    Ticker = ticker,
                    WindowStartUtc = DateTime.UtcNow.AddHours(-24),
                    MessageCount = mentions.Count,
                    HypeLexiconDensity = hypeDensity,
                    NewAccountRatio = newAccountRatio
                };
            }
            catch (Exception ex)
            {
                _log.LogError(ex, "Failed to fetch Reddit data for {Ticker}", ticker);
                return null;
            }
        }

        // =======================
        // OAuth2 Authentication
        // =======================
        private async Task EnsureAuthenticatedAsync()
        {
            if (_accessToken != null && DateTime.UtcNow < _tokenExpiry)
                return;

            await _authLock.WaitAsync();
            try
            {
                // Double-check after acquiring lock
                if (_accessToken != null && DateTime.UtcNow < _tokenExpiry)
                    return;

                _log.LogInformation("Authenticating with Reddit API");

                var authString = Convert.ToBase64String(
                    Encoding.UTF8.GetBytes($"{_config.ClientId}:{_config.ClientSecret}")
                );

                var request = new HttpRequestMessage(HttpMethod.Post, "https://www.reddit.com/api/v1/access_token");
                request.Headers.Authorization = new AuthenticationHeaderValue("Basic", authString);
                request.Headers.UserAgent.ParseAdd(_config.UserAgent);
                request.Content = new FormUrlEncodedContent(new[]
                {
                    new KeyValuePair<string, string>("grant_type", "client_credentials")
                });

                var response = await _http.SendAsync(request);
                response.EnsureSuccessStatusCode();

                var json = await response.Content.ReadAsStringAsync();
                var authResp = JsonSerializer.Deserialize<RedditAuthResponse>(json);

                if (authResp == null || string.IsNullOrEmpty(authResp.access_token))
                    throw new InvalidOperationException("Failed to get access token from Reddit");

                _accessToken = authResp.access_token;
                _tokenExpiry = DateTime.UtcNow.AddSeconds(authResp.expires_in - 60); // 60s buffer

                _log.LogInformation("Reddit authentication successful, token expires at {Expiry}", _tokenExpiry);
            }
            finally
            {
                _authLock.Release();
            }
        }

        // =======================
        // Search Subreddit
        // =======================
        private async Task<List<RedditPost>> SearchSubredditAsync(string subreddit, string ticker)
        {
            var posts = new List<RedditPost>();

            // Search query: ticker symbol (with $ prefix common on Reddit)
            var query = $"${ticker} OR {ticker}";
            var url = $"https://oauth.reddit.com/r/{subreddit}/search?q={Uri.EscapeDataString(query)}&restrict_sr=1&sort=new&t=day&limit={_config.SearchLimitPerSubreddit}";

            var request = new HttpRequestMessage(HttpMethod.Get, url);
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _accessToken);
            request.Headers.UserAgent.ParseAdd(_config.UserAgent);

            var response = await _http.SendAsync(request);
            
            if (!response.IsSuccessStatusCode)
            {
                _log.LogWarning("Reddit search failed for r/{Subreddit}: {Status}", subreddit, response.StatusCode);
                return posts;
            }

            var json = await response.Content.ReadAsStringAsync();
            using var doc = JsonDocument.Parse(json);

            if (!doc.RootElement.TryGetProperty("data", out var data) ||
                !data.TryGetProperty("children", out var children))
            {
                return posts;
            }

            foreach (var child in children.EnumerateArray())
            {
                if (!child.TryGetProperty("data", out var postData))
                    continue;

                var post = new RedditPost
                {
                    Id = postData.GetProperty("id").GetString() ?? "",
                    Author = postData.GetProperty("author").GetString() ?? "",
                    Title = postData.GetProperty("title").GetString() ?? "",
                    Selftext = postData.TryGetProperty("selftext", out var st) ? st.GetString() ?? "" : "",
                    Created = postData.GetProperty("created_utc").GetInt64(),
                    Score = postData.GetProperty("score").GetInt32(),
                    NumComments = postData.GetProperty("num_comments").GetInt32(),
                    Subreddit = subreddit
                };

                // Verify ticker is actually mentioned (Reddit search can be loose)
                var content = (post.Title + " " + post.Selftext).ToUpperInvariant();
                if (content.Contains(ticker.ToUpperInvariant()) || content.Contains($"${ticker.ToUpperInvariant()}"))
                {
                    posts.Add(post);
                }
            }

            _log.LogInformation("Found {Count} posts mentioning {Ticker} in r/{Subreddit}", 
                posts.Count, ticker, subreddit);

            return posts;
        }

        // =======================
        // Get Author Details
        // =======================
        private async Task<List<RedditUser>> GetAuthorDetailsAsync(IEnumerable<string> authors)
        {
            var users = new List<RedditUser>();

            foreach (var author in authors)
            {
                try
                {
                    // Skip deleted/suspended accounts
                    if (author.StartsWith("[deleted]", StringComparison.OrdinalIgnoreCase) ||
                        author.StartsWith("[removed]", StringComparison.OrdinalIgnoreCase))
                        continue;

                    var url = $"https://oauth.reddit.com/user/{author}/about";
                    var request = new HttpRequestMessage(HttpMethod.Get, url);
                    request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _accessToken);
                    request.Headers.UserAgent.ParseAdd(_config.UserAgent);

                    var response = await _http.SendAsync(request);
                    if (!response.IsSuccessStatusCode)
                    {
                        _log.LogWarning("Failed to get details for u/{Author}: {Status}", author, response.StatusCode);
                        continue;
                    }

                    var json = await response.Content.ReadAsStringAsync();
                    using var doc = JsonDocument.Parse(json);

                    if (!doc.RootElement.TryGetProperty("data", out var data))
                        continue;

                    users.Add(new RedditUser
                    {
                        Name = author,
                        Created = data.GetProperty("created_utc").GetInt64(),
                        LinkKarma = data.TryGetProperty("link_karma", out var lk) ? lk.GetInt32() : 0,
                        CommentKarma = data.TryGetProperty("comment_karma", out var ck) ? ck.GetInt32() : 0
                    });

                    // Rate limiting (60 req/min)
                    await Task.Delay(TimeSpan.FromSeconds(1));
                }
                catch (Exception ex)
                {
                    _log.LogWarning(ex, "Error fetching details for u/{Author}", author);
                }
            }

            return users;
        }

        // =======================
        // Analysis Helpers
        // =======================
        private bool ContainsHypeLanguage(string text)
        {
            var lower = text.ToLowerInvariant();
            return _hypeTerms.Any(term => lower.Contains(term.ToLowerInvariant()));
        }

        private bool IsNewAccount(RedditUser user, int dayThreshold)
        {
            var accountAge = DateTime.UtcNow - DateTimeOffset.FromUnixTimeSeconds(user.Created);
            return accountAge.TotalDays < dayThreshold;
        }

        public void Dispose()
        {
            _authLock?.Dispose();
        }
    }

    // =======================
    // Factory/Configuration Helper
    // =======================
    public static class RedditServiceExtensions
    {
        public static IServiceCollection AddRedditClient(this IServiceCollection services)
        {
            // Get config from environment variables
            var clientId = Environment.GetEnvironmentVariable("REDDIT_CLIENT_ID");
            var clientSecret = Environment.GetEnvironmentVariable("REDDIT_CLIENT_SECRET");
            
            if (string.IsNullOrEmpty(clientId) || string.IsNullOrEmpty(clientSecret))
            {
                // Fall back to stub if credentials not configured
                services.AddSingleton<ISocialClient, SocialClientStub>();
                return services;
            }

            var config = new RedditConfig
            {
                ClientId = clientId,
                ClientSecret = clientSecret,
                UserAgent = "VeritasAlpha/1.0",
                Subreddits = (Environment.GetEnvironmentVariable("REDDIT_SUBREDDITS") ?? "stocks,investing")
                    .Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries)
            };

            services.AddSingleton(config);
            services.AddHttpClient<ISocialClient, RedditClient>();

            return services;
        }
    }
}